### ISSUE TITLE
Lab 5: (Test case 8) Time Limit Exceeded

### ISSUE THREAD

**Issue by airrehn** [nus-cs2030/2223-s1#190]
<img width="311" alt="Screenshot 2022-10-13 at 12 47 31 AM" src="https://user-images.githubusercontent.com/110771839/195401559-07e0b984-6d73-425b-bd06-37ce8fc246de.png">

Hi! I need some help for Lab 5. 

May I know what is _Test Case 8_? For this particular test case, my code does not print anything and the error shows "Time Limit Exceeded" and not "Incorrect Output". May I know what could be causing this problem?

Thank you!

Edit: 
To give some context, the logic for the events is that a `Wait` could return another `Wait` or `Serve` depending on the number of customers in queue. However, I think somewhere down the line the PQ is in an infinite loop where a Wait will keep returning another Wait, not allowing the PQ to proceed. 

Anyone who had used a similar logic, mind sharing what was your condition for letting the `Wait` return another `Wait`? 

Mine was to check if the customer queue currently had 2 or more in it. (I only deduct the queue number in the `Serve` event's `nextEvent()` method ) I see how this could be a problem if a `Wait` event is in queue with a `Serve` event that had the same timing and a later customer number. 

Any help is appreciated thank you!

**Comment by Y-ZR** [comment]
Hello, time limit exceeded could mean there was no output from your program after a set amount of time by cc.

**Comment by bransometan** [comment]
Another problem that I could think of is that your queue somehow ended up in an infinite loop, resulting in time limit exceeded. It would be good to check if your events are updated correctly into the queue or if your servers service time are updated correctly.

**Comment by airrehn** [comment]
> Another problem that I could think of is that your queue somehow ended up in an infinite loop, resulting in time limit exceeded. It would be good to check if your events are updated correctly into the queue or if your servers service time are updated correctly.

**Comment by airrehn** [comment]
> Another problem that I could think of is that your queue somehow ended up in an infinite loop, resulting in time limit exceeded. It would be good to check if your events are updated correctly into the queue or if your servers service time are updated correctly.



> Another problem that I could think of is that your queue somehow ended up in an infinite loop, resulting in time limit exceeded. It would be good to check if your events are updated correctly into the queue or if your servers service time are updated correctly.

Yeah! I think the issue is that my Wait returns another Wait and with some specific condition of other Events in the PQ, the PQ cannot proceed. :/

**Comment by keefetok** [comment]
Hi! I had the same problem as you before. I realised it was easier to use another class as a Wait to limit my wait’s toString to just a wait event, while i manipulate my other wait class to not print anything but store a pair of objects.

**Comment by airrehn** [comment]
> Hi! I had the same problem as you before. I realised it was easier to use another class as a Wait to limit my wait’s toString to just a wait event, while i manipulate my other wait class to not print anything but store a pair of objects.

Hi! To manipulate my Wait class to not print anything, I just added an additional boolean property (toPrint) to all my Events and if I get a Wait from a Wait, I'll force it to be false. 

Also, I don't know how to replicate the inputs of Test case 8 that would give me this problem hahaha.

You mentioned you had this same problem before and it was easier to use another Class as a Wait haha. Did you mean you returned another Event from the Wait? Thank you!

**Comment by keefetok** [comment]
Ya… to make things clearer: 

Arrive > Serve > Done
Arrive > Wait > Can Serve > Done
Arrive > Wait > Wait Again (another wait) > Can Serve > Done


So as you can see, it’s an intermediate class in between wait and being served.

**Comment by keefetok** [comment]
oh ya, if u wna debug this particular test case, i suggest to use the set of arrival and services given in the lab notes (last week’s) to debug. Do change the service times to 0.4 though, rather than 4

**Comment by nixonwidjaja** [comment]
Try to check if your while (!pq.isEmpty()) triggers an infinite loop

**Comment by airrehn** [comment]
> Ya… to make things clearer:
> 
> Arrive > Serve > Done Arrive > Wait > Can Serve > Done Arrive > Wait > Wait Again (another wait) > Can Serve > Done _edit_: just gave your problem a greater read, and my implementation can be found in the scenario ^, rather than using the people in queue, i use the above to check whether i can serve/wait
> 
> So as you can see, it’s an intermediate class in between wait and being served.
> 
> Doing another wait works, and you have found a solution to prevent a print, but i think your times are being messed up & that’s why it’s returning an infinite loop

Hi! I woke up, decided to change the condition to if `canServe` as you mentioned and it worked! Thank you so much. Hope this may help others too (:

**Comment by qwqdrz** [comment]
I think it may be possibly caused by an infinite loop. A good way to see what goes wrong is to add some System.out.println() lines in the code and check with ur own example first in vim to see if the code can run in the way you expected.

**Comment by wjiayis** [comment]
> Arrive > Serve > Done Arrive > Wait > Can Serve > Done Arrive > Wait > Wait Again (another wait) > Can Serve > Done _edit_: just gave your problem a greater read, and my implementation can be found in the scenario ^, rather than using the people in queue, i use the above to check whether i can serve/wait
> 
> So as you can see, it’s an intermediate class in between wait and being served.
> 
> Doing another wait works, and you have found a solution to prevent a print, but i think your times are being messed up & that’s why it’s returning an infinite loop

While this works fine, I think a cleaner implementation would be to use Math.max() when determining the Serve event's time!
